### DISCUSSION TITLE
BudiBase not Showing Data correctly for linked Tables after Importing Big Chunk of Data

### DISCUSSION THREAD

**Discussion by tinomerl** [Budibase/budibase#11209] [Help] [answered]
**Preface**
I already updated to the latest version and am now on 2.8.10. I already searched the Issues, Discussion and Google.

**Setup**
I'm running BudiBase as a Single Container in an Azure Container Instance, according to [this tutorial](https://docs.budibase.com/docs/azure-container-instances). My Database is a PostgreSQL Flexible Server also on Azure.

I'm trying to setup up BudiBase as a way to make our workflow for data entry easier. Therefore i created a new database in a normalized way, setup BudiBase with test Data and I'm trying now to import the Data. The main use Case will be entry of Ad Spendings, which can't be automatically fetched via API. I import the data via a Python Script into the Database directly

**Problem Description**
The central table is ad_spendings and contains about 27k rows. I normalized the campaigns, country and source. In the central Ad table they are present as foreign Keys. Normally we have about 35 different countries, 175 sources and 63 campaigns. They are all displayed correctly until i import the data for the Ad Spendings into the Database. Afterwards only 2 values of either campaigns, source and country will be displayed in the BudiBase Backend and on the Screens for these Tables. If I go to the Ad Spendings Screen, I can see the Name of the missing Values displayed correctly. If i click on the entry however to edit it, the Relationship Picker will display "Null" and i can now only select from the 2 which Budibase will show me. The problem seems not to be with the database and the data, as all relations in the database are displayed correctly and all columns have appropriate Foreign Keys.

**What I Tried**
- Dropping the whole Database and Set it up as fresh
- Importing only chunks of the ad spendings
  - When i do this the data will be displayed correctly until i cross a certain threshold of about 100 Rows
  - But it seems that when i enter the data manually the error does not persist
- Import the data via BudiBase Backend as a csv
  - It will not recognize my foreign Keys
- Deleting the Relationship in BudiBase and declaring it again
  - After deleting the Relationship the data will display correct in the e.g. Country Table. But after setting up the Relationship again the Error is there again
- Setting up a new App with the Database.
- Importing Data via the BudiBase API and the Create Row method
  - Error also happens after a about hundred rows

**Comment by melohagan** [comment]
Hey @tinomerl 

Would you be able to include some screenshots to outline the problem as well please?

**Comment by tinomerl** [comment]
Hey @melohagan,

thanks for reaching out. Sure.

__Database schema__
![postgres - public](https://github.com/Budibase/budibase/assets/35485536/10cdeb1b-dcbd-4331-8eaf-1a535937623c)

# Before Importing Ad Spendings

__Country Table__
_In Budibase_
![image](https://github.com/Budibase/budibase/assets/35485536/7570483d-24e7-4d43-aff4-5e03e0c4a9d5)
_In Database_
![image](https://github.com/Budibase/budibase/assets/35485536/cecf813e-3ef1-480d-aac4-559e057e8eab)

__Source Table__
_In Budibase_
![image](https://github.com/Budibase/budibase/assets/35485536/9b3dbdca-447f-4f07-bb7d-c52d971a1517)
_In Database_
![image](https://github.com/Budibase/budibase/assets/35485536/ad783e91-beb2-49de-9e20-491a8d031148)

# After Importing Ad Spendings

__Country Table__
_In Budibase_
<img width="1214" alt="image" src="https://github.com/Budibase/budibase/assets/35485536/73f7b442-59b0-44f2-9044-e37dba13a78f">
_In Database_
<img width="340" alt="image" src="https://github.com/Budibase/budibase/assets/35485536/fbcbb23a-737c-4471-a251-7243730e2fba">

__Source Table__
_In Budibase_
<img width="666" alt="image" src="https://github.com/Budibase/budibase/assets/35485536/da21a8d1-69bf-474c-a510-25832953b806">
_In Database_
<img width="315" alt="image" src="https://github.com/Budibase/budibase/assets/35485536/f3e6a4f9-12a6-4f59-a3b1-a665fe528b9f">

__Number of Rows in Ad Spendings__
<img width="180" alt="image" src="https://github.com/Budibase/budibase/assets/35485536/b8fb1207-324a-4da2-b83d-11805d1c98d8">

__Taken from Ad Spendings Screen__
![image](https://github.com/Budibase/budibase/assets/35485536/3a2c09d7-9b3a-4028-9883-8834cc15b54b)

__When i click on one of the Entries to edit it__
![image](https://github.com/Budibase/budibase/assets/35485536/bbc068bb-8653-498b-96a8-a2aec33fd1fa)

_These are not all, there are still some missing_
![image](https://github.com/Budibase/budibase/assets/35485536/074e90b4-02cb-45b7-8241-b9f69bcb5474)



Does this help? Do you need anything else?

**Reply by tinomerl** [reply]
@melohagan do you have any updates on this?

**Reply by melohagan** [reply]
Hey @tinomerl 

Sorry for the late reply, but I'm pretty sure the issue you are seeing is the same as this one: https://github.com/Budibase/budibase/issues/11386

Budibase does LEFT JOINs when handling relationships, which means if you have more than one relationship per table, the number of rows being queried can quickly expand.

For SQL databases there is a limit of 5000 rows, however you could take a look at the `SQL_MAX_ROWS` environment variable: https://docs.budibase.com/docs/hosting-settings

With that said, I would recommend using custom queries instead to handle the JOINs yourself if you are running into this problem, as the performance will be better.

**Reply by tinomerl** [reply]
Hey @melohagan, 

thanks for following up! So just to make sure i get this right. Instead of using the table itself as a datasource for my view i should create a custom query for reading, updating and creating rows each right?

**Reply by melohagan** [reply]
Yes that's right. The downside is that you won't be able to make use of the relationship pickers, you'll have to use options pickers with data providers instead.

https://docs.budibase.com/docs/option-picker#using-a-data-provider

**Reply by tinomerl** [reply, marked as answer]
Thanks again. I figured it out now. I managed to make it work with just one custom query, for reading only.  This one is used for the main Table. The option Pickers gave me a headache but i made it work as [described here](https://github.com/Budibase/budibase/discussions/9449#discussioncomment-4822943). So whenever i update a row or create a new one, i just use the regular ad_spendings table for the form and hide the columns for foreign Keys. In this Form i have as many Data Providers and Option Pickers as there are Foreign Keys. The option Pickers have an on Change Action, that updates these foreign Key Fields.
